cmd/cluster-replication.go

```diff
@@ -0,0 +1,968 @@
+		if d == globalDeploymentID {
+			err := c.PeerBucketMakeWithVerHandler(ctx, bucket, opts)
+			if err != nil {
+				fmt.Printf("%s: MakeWithVer: %v\n", c.state.Name, err)
+				return wrapCRErr(err)
+			}
+			continue
+		}
+
+		admClient, err := c.getAdminClient(ctx, d)
+		if err != nil {
+			return wrapCRErr(err)
+		}
+
+		err = admClient.CRInternalBucketOps(ctx, bucket, madmin.MakeWithVerBktOp, optsMap)
+		if err != nil {
+			fmt.Printf("%s->%s: MakeWithVer: %v\n", c.state.Name, p.Name, err)
+			return wrapCRErr(err)
+		}
+	}
+
+	// Handle bucket remote and replication addition.
+	for d, p := range c.state.Peers {
+		if d == globalDeploymentID {
+			err := c.PeerBucketConfigureReplHandler(ctx, bucket)
+			if err != nil {
+				fmt.Printf("%s: ConfigureRepl: %v\n", c.state.Name, err)
+				return wrapCRErr(err)
+			}
+			continue
+		}
+
+		admClient, err := c.getAdminClient(ctx, d)
+		if err != nil {
+			return wrapCRErr(err)
+		}
+
+		err = admClient.CRInternalBucketOps(ctx, bucket, madmin.ConfigureReplBktOp, nil)
+		if err != nil {
+			fmt.Printf("%s->%s: ConfigureRepl: %v\n", c.state.Name, p.Name, err)
+			return wrapCRErr(err)
+		}
+	}
+	return nil
+}
+
+// DeleteBucketHook - called during a regular delete bucket call when cluster
+// replication is enabled. It is responsible for the deletion of the same bucket
+// on remote clusters.
+func (c *ClusterReplMgr) DeleteBucketHook(ctx context.Context, bucket string, forceDelete bool) error {
+	// At this point, the local bucket is deleted.
+
+	c.RLock()
+	defer c.RUnlock()
+	if !c.Enabled {
+		return nil
+	}
+
+	op := madmin.DeleteBucketBktOp
+	if forceDelete {
+		op = madmin.ForceDeleteBucketBktOp
+	}
+
+	// Send bucket delete to other clusters.
+	for d, p := range c.state.Peers {
+		if d == globalDeploymentID {
+			continue
+		}
+
+		admClient, err := c.getAdminClient(ctx, d)
+		if err != nil {
+			return wrapCRErr(err)
+		}
+
+		err = admClient.CRInternalBucketOps(ctx, bucket, op, nil)
+		if err != nil {
+			fmt.Printf("%s->%s: DeleteBucket: %v\n", c.state.Name, p.Name, err)
+			return wrapCRErr(err)
+		}
+	}
+	return nil
+}
+
+// PeerBucketMakeWithVerHandler - creates bucket and enables versioning.
+func (c *ClusterReplMgr) PeerBucketMakeWithVerHandler(ctx context.Context, bucket string, opts BucketOptions) error {
+	objAPI := newObjectLayerFn()
+	err := objAPI.MakeBucketWithLocation(ctx, bucket, opts)
+	if err != nil {
+		// Check if this is a bucket exists error.
+		_, ok1 := err.(BucketExists)
+		_, ok2 := err.(BucketAlreadyExists)
+		if !ok1 && !ok2 {
+			fmt.Printf("%s: isFromPeer - mb err: %#v\n", c.state.Name, err)
+			return wrapCRErr(err)
+		}
+	} else {
+		// Load updated bucket metadata into memory as new
+		// bucket was created.
+		globalNotificationSys.LoadBucketMetadata(GlobalContext, bucket)
+		// fmt.Printf("%s: make bucket %s success!\n", c.state.Name, bucket)
+	}
+
+	// Enable versioning on the bucket.
+	config, err := globalBucketVersioningSys.Get(bucket)
```

Comment:
For site replication to work, we implicitly enable versioning even if the user-issued make-bucket call did not specify versioning. We do need this code.

Reply:
Correct - but why not set opts.VersioningEnabled in MakeBucketHook as this hook will be invoked only for cluster replication, and let MakeBucketWithLocation enable versioning as part of bucket creation.

Reply:
MakeBucketWithLocation does not seem to handle enabling versioning - https://github.com/minio/minio/blob/master/cmd/erasure-server-pool.go#L598 - likely because there is no way to enable versioning in the call to make-bucket from a user-client. I didn't test it - would be good if you could confirm.

Reply:
oh wait - you are right, versioning is enabled  only if object locking is specified.